### ISSUE TITLE
Parse function configuration file client side

### ISSUE THREAD

**Issue by sebgoa** [triggermesh/tm#32]
`tm` should be able to parse a `serverless.yml` like file that defines the multiple functions a repo, their arguments, runtime etc.

then deploy services for each function

**Comment by tzununbekov** [comment]
In `serverless` branch I pushed several changes that makes possible to deploy multiple functions with `--definition <path>` parameter, but deployment options are very limited now. Here is an example of possible serverless.yaml:

```
service: serverless-test

provider:
  name: triggermesh
  runtime: openfaas-go-runtime

functions:
  hello:
    handler: foo/main.go
  world:
    handler: bar/main.go
``` 


While I was working on that, few questions appeared:
1) Do we need to strictly keep to serverless framework yaml file format and its deployment options?
2) Are we planning to fully support applications that run on serverless framework?
Depending on answers, I'll probably have more questions. 
One of the main concern here is runtimes that we don't support now (aws-lambda for example) and eventing

**Comment by sebgoa** [comment]
nice.

At a high level:

1. We will only support functions that deploy to knative/triggermesh so no need to worry about other targets like AWS.

2. Let's skip eventing in this first iteration

3. If you do `tm deploy` it should by default look for a `serverless.yml` and deploy that. if it does not find it then print an error message. if you specify it at the cli do `tm deploy -f foo.yaml`

Now on to details:

1. The runtime should be a name referencing to an existing build template. Or a URL to a build template. If it is a URL then `tm` should create it. Make sure that you can also set the runtime per function and not only at the provider level.

2. Add memorySize like:

```
provider:
  name: knative
  runtime: https://raw.githubusercontent.com/triggermesh/openfaas-runtime/master/go/openfaas-go-runtime.yaml
  memorySize: 512 
```
This should set a resources limit in the Configuration spec...which becomes a resource limit in the underlying Pod.

3. Add 'description' field for functions. This should become an annotation set in the metadata

4. Add 'labels' which become labels in the metadata

5. Add 'environment' which is a list of environment variables.

see: https://github.com/serverless/examples/blob/master/aws-node-env-variables/serverless.yml

6. Add support for multiple functions in an array like

```
# Multiple functions

functions:
  - ${file(../foo-functions.yml)}
  - ${file(../bar-functions.yml)}
```

**Comment by tzununbekov** [comment]
Update in `serverless` branch:
- `tm deploy` is looking for serverless.yaml by default. `tm deploy -f <path>` accepts local or remote path to serverless.yaml
- both per file and per function definition of runtimes and env variables
- `tm deploy` accepts buildtemplate local or remote path to deploy it
- no need to specify source path type in tm deployment commands anymore, eg:
```
tm deploy service foo --from-source <url>
```
```
tm deploy service foo --from-image <registry>
```
can be replaced with 
```
tm deploy service foo -s <path>
```    
but old format is kept for back compatibility 

Todo:
- description, labels and multiple functions should be an easy thing to add 
- memory limitation - there was knative [limitation](https://github.com/triggermesh/tm/issues/6#issuecomment-422371456) about pod resources management, need to review newer version of knative 

- test everything: this branch already has many changes so before merge I'd test all deployment flows not to break anything

**Comment by tzununbekov** [comment]
I think I'm done with this feature for now; everything merged into master and new tag (0.0.5) has been pushed
Container's resource management is still knative controller responsibility and user can't change it: https://github.com/triggermesh/tm/issues/6#issuecomment-422371456

Repository Readme updated with new CLI features.   

I want to highlight really nice option to combine different repositories by specifying URL as include value:  
https://github.com/tzununbekov/serverless/blob/master/serverless.yaml#L32

Let's track any other requests related to serverless.yaml parsing in separate issues
